File: app/src/main/java/com/sevtinge/cemiuiler/module/systemui/statusbar/StatusBarIcon.java
package com.sevtinge.cemiuiler.module.systemui.statusbar;

import com.sevtinge.cemiuiler.module.base.BaseHook;

public class StatusBarIcon extends BaseHook {

    @Override
    public void init() {

        MethodHook mIconHook = new MethodHook() {
            @Override
            protected void before(MethodHookParam param) {
                String iconType = (String) param.args[0];
                switch (checkSlot(iconType)) {
                    case 1 -> param.args[1] = true;
                    case 2 -> param.args[1] = false;
                }
            }
        };

        findAndHookMethod("com.android.systemui.statusbar.phone.StatusBarIconControllerImpl", "setIconVisibility", String.class, boolean.class, mIconHook);
        findAndHookMethod("com.android.systemui.statusbar.phone.MiuiDripLeftStatusBarIconControllerImpl", "setIconVisibility", String.class, boolean.class, mIconHook);
    }


    private static int checkSlot(String slotName) {
        int vpn = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_vpn", 0);
        int alarmClock = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_alarm_clock", 0);
        int nfc = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_nfc", 0);
        int zen = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_zen", 0);
        int volume = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_volume", 0);
        int wifi = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_wifi", 0);
        int wifi_slave = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_wifi_slave", 0);
        int airplane = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_airplane", 0);
        int location = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_location", 0);
        int hotspot = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_hotspot", 0);
        int headset = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_headset", 0);

        return switch (slotName) {
            case "vpn" ->// vpn
                isEnable(vpn) ? vpn : 0;
            case "alarm_clock" ->// 闹钟
                isEnable(alarmClock) ? alarmClock : 0;
            case "nfc" ->// nfc
                isEnable(nfc) ? nfc : 0;
            case "zen" ->// 勿扰模式
                isEnable(zen) ? zen : 0;
            case "volume" ->// 声音
                isEnable(volume) ? volume : 0;
            case "wifi" ->// wifi
                isEnable(wifi) ? wifi : 0;
            case "wifi_slave" ->// 辅助wifi
                isEnable(wifi_slave) ? wifi_slave : 0;
            case "airplane" ->// 飞行模式
                isEnable(airplane) ? airplane : 0;
            case "location" ->// 位置信息
                isEnable(location) ? location : 0;
            case "hotspot" ->// 热点
                isEnable(hotspot) ? hotspot : 0;
            case "headset" ->// 耳机
                isEnable(headset) ? headset : 0;
            default -> 0;
        };
    }

    private static boolean isEnable(int i) {
        return 0 < i && i < 3;
    }
}

File: app/src/main/java/com/sevtinge/cemiuiler/module/hook/systemui/statusbar/icon/all/StatusBarIcon.kt
package com.sevtinge.cemiuiler.module.hook.systemui.statusbar.icon.all

import com.github.kyuubiran.ezxhelper.ClassUtils.loadClass
import com.github.kyuubiran.ezxhelper.HookFactory.`-Static`.createHook
import com.github.kyuubiran.ezxhelper.MemberExtensions.paramCount
import com.github.kyuubiran.ezxhelper.finders.MethodFinder.`-Static`.methodFinder
import com.sevtinge.cemiuiler.module.base.BaseHook
import com.sevtinge.cemiuiler.utils.devicesdk.isAndroidU

class StatusBarIcon : BaseHook() {
    override fun init() {
        loadClass("com.android.systemui.statusbar.phone.StatusBarIconControllerImpl").methodFinder().first {
            name == "setIconVisibility" && paramCount == 2
        }.createHook {
            before { param ->
                val iconType = param.args[0] as String
                when (checkSlot(iconType)) {
                    1 -> param.args[1] = true
                    2 -> param.args[1] = false
                }
            }
        }


        loadClass(if (isAndroidU())
            "com.android.systemui.statusbar.phone.StatusBarIconControllerImpl"
        else
            "com.android.systemui.statusbar.phone.MiuiDripLeftStatusBarIconControllerImpl").methodFinder().first {
            name == "setIconVisibility" && paramCount == 2
        }.createHook {
            before { param ->
                val iconType = param.args[0] as String
                when (checkSlot(iconType)) {
                    1 -> param.args[1] = true
                    2 -> param.args[1] = false
                }
            }
        }
    }

    companion object {
        private fun checkSlot(slotName: String): Int {
            val vpn = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_vpn", 0)
            val alarmClock = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_alarm_clock", 0)
            val nfc = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_nfc", 0)
            val zen = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_zen", 0)
            val volume = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_volume", 0)
            val wifi = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_wifi", 0)
            val wifi_slave = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_wifi_slave", 0)
            val airplane = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_airplane", 0)
            val location = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_location", 0)
            val hotspot = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_hotspot", 0)
            val headset = mPrefsMap.getStringAsInt("system_ui_status_bar_icon_headset", 0)
            return when (slotName) {
                "vpn" ->  // vpn
                    if (isEnable(vpn)) vpn else 0

                "alarm_clock" ->  // 闹钟
                    if (isEnable(alarmClock)) alarmClock else 0

                "nfc" ->  // nfc
                    if (isEnable(nfc)) nfc else 0

                "zen" ->  // 勿扰模式
                    if (isEnable(zen)) zen else 0

                "volume" ->  // 声音
                    if (isEnable(volume)) volume else 0

                "wifi" ->  // wifi
                    if (isEnable(wifi)) wifi else 0

                "wifi_slave" ->  // 辅助wifi
                    if (isEnable(wifi_slave)) wifi_slave else 0

                "airplane" ->  // 飞行模式
                    if (isEnable(airplane)) airplane else 0

                "location" ->  // 位置信息
                    if (isEnable(location)) location else 0

                "hotspot" ->  // 热点
                    if (isEnable(hotspot)) hotspot else 0

                "headset" ->  // 耳机
                    if (isEnable(headset)) headset else 0

                else -> 0
            }
        }

        private fun isEnable(i: Int): Boolean {
            return i in 1..2
        }
    }
}
